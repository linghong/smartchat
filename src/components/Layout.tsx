import { useState, useEffect, FC, ReactNode } from 'react';

import Footer from '@/src/components/Footer';
import Header from '@/src/components/Header';
import Sidebar from '@/src/components/Sidebar';
import { OptionType } from '@/src/types/common';

interface LayoutProps {
  children: ReactNode;
  isConfigPanelVisible: boolean;
  setIsConfigPanelVisible: (isConfigPanelVisible: boolean) => void;
  namespacesList: OptionType[] | null;
  chatId: string | null;
  setChatId: (chatId: string | null) => void;
}

const Layout: FC<LayoutProps> = ({
  children,
  isConfigPanelVisible,
  setIsConfigPanelVisible,
  namespacesList,
  chatId,
  setChatId
}) => {
  const [isSidebarOpen, setIsSidebarOpen] = useState(true);
  const [isMobile, setIsMobile] = useState(false);

  useEffect(() => {
    const checkIsMobile = () => {
      setIsMobile(window.innerWidth <= 480);
    };

    // Run on initial mount
    checkIsMobile();

    window.addEventListener('resize', checkIsMobile);

    return () => {
      window.removeEventListener('resize', checkIsMobile);
    };
  }, []);

  return (
    <div className="flex flex-col w-full h-screen">
      <Header
        isSidebarOpen={isSidebarOpen}
        setIsSidebarOpen={setIsSidebarOpen}
        isConfigPanelVisible={isConfigPanelVisible}
        setIsConfigPanelVisible={setIsConfigPanelVisible}
      />
      <div className="flex flex-row w-full h-full">
        {isMobile && isSidebarOpen && (
          <div className="bg-slate-500 text-slate-50 h-full w-full">
            <Sidebar
              setIsSidebarOpen={setIsSidebarOpen}
              namespacesList={namespacesList}
              chatId={chatId}
              setChatId={setChatId}
            />
          </div>
        )}
        {isMobile && !isSidebarOpen && (
          <main className="flex flex-col w-full lg:w-70vw h-full px-2">
            {children}
            <Footer />
          </main>
        )}
        {!isMobile && isSidebarOpen && (
          <div className="flex flex-row w-full h-full">
            <div className="bg-slate-500 text-slate-50 h-full text-md w-full xs:w-56 xl:w-64 2xl:w-72 fixed top-0 left-0 bottom-0 z-10">
              <Sidebar
                setIsSidebarOpen={setIsSidebarOpen}
                namespacesList={namespacesList}
                chatId={chatId}
                setChatId={setChatId}
              />
            </div>
            <div className="flex flex-col w-full ml-56 xl:ml-64 2xl:ml-72">
              <main
                className={`flex flex-col h-full w-full max-w-screen-2xl mx-auto px-2 sm:px-4 lg:px-8`}
              >
                <div className="flex-grow">{children}</div>
                <Footer />
              </main>
            </div>
          </div>
        )}
        {!isMobile && !isSidebarOpen && (
          <main className="flex flex-col h-full w-full lg:w-70vw px-2 mx-auto">
            {children}
            <Footer />
          </main>
        )}
      </div>
    </div>
  );
};

export default Layout;
