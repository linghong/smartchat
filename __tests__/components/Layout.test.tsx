import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';

import Layout from '@/src/components/Layout';

// Mock Header component, default sidebar is open
// When Header is clicked, the sidebar will be closed
jest.mock('@/src/components/Header', () => {
  return {
    __esModule: true,
    default: ({
      setIsSidebarOpen
    }: {
      setIsSidebarOpen: (isOpen: boolean) => void;
    }) => <div onClick={() => setIsSidebarOpen(false)}>Header Mock</div>
  };
});

// Mock Sidebar component
jest.mock('@/src/components/Sidebar', () => {
  return {
    __esModule: true,
    default: () => <div>Sidebar Mock</div>
  };
});

// Mock Footer component
jest.mock('@/src/components/Footer', () => {
  return {
    __esModule: true,
    default: () => <div>Footer Mock</div>
  };
});

describe('Layout Component', () => {
  beforeEach(() => {
    // Reset the viewport size before each test
    global.innerWidth = 1024;
    global.dispatchEvent(new Event('resize'));
  });

  it('renders Header, Sidebar (conditionally), and Footer correctly on initial load', () => {
    const messageSubjectList = ['Test Subject'];
    const { getByText } = render(
      <Layout messageSubjectList={messageSubjectList}>
        <div>Child Content</div>
      </Layout>
    );

    // Check for Header and Footer
    expect(getByText('Header Mock')).toBeInTheDocument();
    expect(getByText('Footer Mock')).toBeInTheDocument();

    // Sidebar should be visible by default in desktop mode (mock as desktop initially if not specified)
    expect(getByText('Sidebar Mock')).toBeInTheDocument();

    // Check that child content is rendered
    expect(getByText('Child Content')).toBeInTheDocument();
  });

  it('does not render Sidebar when it is in mobile mode and Header menu is clicked to be closed', () => {
    const messageSubjectList = ['Test Subject'];
    const { getByText, queryByText } = render(
      <Layout messageSubjectList={messageSubjectList}>
        <div>Child Content</div>
      </Layout>
    );

    // Simulate mobile view
    act(() => {
      global.innerWidth = 480;
      global.dispatchEvent(new Event('resize'));
    });

    act(() => {
      fireEvent.click(getByText('Header Mock'));
    });

    expect(queryByText('Sidebar Mock')).not.toBeInTheDocument();
    expect(getByText('Child Content')).toBeInTheDocument();
  });

  it('renders Sidebar in mobile mode when sidebar is open', () => {
    const messageSubjectList = ['Test Subject'];
    const { getByText, queryByText } = render(
      <Layout messageSubjectList={messageSubjectList}>
        <div>Child Content</div>
      </Layout>
    );

    act(() => {
      global.innerWidth = 480;
      global.dispatchEvent(new Event('resize'));
    });

    expect(getByText('Sidebar Mock')).toBeInTheDocument();
    expect(queryByText('Child Content')).not.toBeInTheDocument();
  });
});
