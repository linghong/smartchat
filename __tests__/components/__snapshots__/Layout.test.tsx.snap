// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Component should match Layout component snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col w-full h-screen"
  >
    <div>
      Header Mock
    </div>
    <div
      class="flex flex-row w-full h-full"
    >
      <div
        class="flex flex-row w-full h-full"
      >
        <div
          class="bg-slate-500 text-slate-50 h-full text-md w-full xs:w-56 xl:w-64 2xl:w-72 fixed top-0 left-0 bottom-0 z-10"
        >
          <div>
            Sidebar Mock
          </div>
        </div>
        <div
          class="flex flex-col w-full ml-56 xl:ml-64 2xl:ml-72"
        >
          <main
            class="flex flex-col h-full w-full max-w-screen-2xl mx-auto px-2 sm:px-4 lg:px-8"
          >
            <div
              class="flex-grow"
            >
              <div>
                Child Content
              </div>
            </div>
            <div>
              Footer Mock
            </div>
          </main>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
